aeppl/joint_logprob.py

```diff
@@ -44,25 +44,34 @@ def factorized_joint_logprob(
         \sigma^2 \sim& \operatorname{InvGamma}(0.5, 0.5) \\
         Y \sim& \operatorname{N}(0, \sigma^2)
 
-    If we create a value variable for ``Y_rv``, i.e. ``y_vv = at.scalar("y")``,
-    the graph of ``factorized_joint_logprob({Y_rv: y_vv})`` is equivalent to the
-    conditional probability :math:`\log p(Y = y \mid \sigma^2)`, with a stochastic
-    ``sigma2_rv``. If we specify a value variable for ``sigma2_rv``, i.e.
-    ``s_vv = at.scalar("s2")``, then ``factorized_joint_logprob({Y_rv: y_vv, sigma2_rv: s_vv})``
-    yields the joint log-probability of the two variables.
+    The graph generated by ``conditional_logprob(Y_rv)`` corresponds to the
+    conditional log-probability :math:`P(Y=y \mid \sigma^2)` where
+    :math:`\sigma^2` is a stochastic variables. The graphs generated by
+    ``conditional_logprob(sigma2_rv, Y_rv)`` represent :math:`\sigma^2` and
+    :math:`Y`'s respective conditional log-probabilities, :math:`\log P(\sigma^2 = s)`
+    and :math:`\log p(Y = y | \sigma^2 = s)`.
 
-    .. math::
+    `conditional_logprob` generates the value variables that correspond to the
+    measurable variables for which it produces a conditional log-probability
+    graph and returns them along with the graphs:
 
-        \log p(Y = y, \sigma^2 = s) =
-            \log p(Y = y \mid \sigma^2 = s) + \log p(\sigma^2 = s)
+    .. code-block:: python
+
+        import aesara.tensor as at
+
+        sigma2_rv = at.random.invgamma(0.5, 0.5)
+        Y_rv = at.random.normal(0, at.sqrt(sigma2_rv))
+
+        logprobs = conditional_logprob(Y_rv, sigma2_rv)
+        # print(logprobs.keys())
+        # [sigma2_vv, Y_vv]
```

Comment:
Would we want the keys to be the generated value variables?  A caller will generally need to match the original random variable arguments with their respective results.  We could instead return a map from the original random variable arguments to their results.  We could also use the call argument order, though.

Reply:
It is indeed generally more convenient for callers to use random variables as keys in `logprobs`, I will make this change.

I also entertained the idea of implementing the following interface, where both maps are indexed by the random variables:

```python
# Random variables are passed as *args
logprobs, value_variables = conditional_logprob(Y_rv, sigma2_rv)
print(logprobs.keys())
# [Y_rv, sigma2_rv]

print(value_variables)
# {Y_rv: Y_vv, sigma2_rv: sigma2_vv}
```

This undoubtedly would be very convenient. However, there are cases in which we want to use value variables that contain data. The simplest solution would be to pass them as keyword arguments, but it does get clunky where also passing e.g. `extra_rewrites`:

```python
logprobs, value_variables = conditional_logprob(
    sigma2_rv,
    Y_rv=Y_data,
    extra_rewrites=rewrites
)
```

My opinion is that whatever is passed via `extra_rewrites` belongs to a separate function, but before (if) we do that we need to compose with it. Non-ambiguous alternatives:

```python
logprobs, value_variables = conditional_logprob(
    sigma2_rv,
    observed={Y_rv: Y_data},
    extra_rewrites=rewrites
)
print(value_variables)
# {sigma2_rv: sigma2_vv}

logprobs, value_variables = aeppl.conditional_logprob(sigma2_rv, Y_rv)
logprobs = aeppl.observe(logprobs, {Y_rv: Y_data})
```

But what is the difference between this and passing `Y_data` to the compiled function, really? I lean towards leaving "observation" considerations out of `AePPL` (it has a high-level wrapper feel to it) and go with `conditional_logprob(Y_rv, sigma2_rv)`. I really like the `conditional_logprob({x_rv: x_vv, Y_rv: Y_data})` interface but having to clone the random variables by hand quickly becomes tedious.

In short, there's no clear winner. We can readily eliminate the ambiguous option, but besides that it's all about the tradeoffs we're ready to make.

Reply:
Anyway the `observed=` interface is trivial enough to implement that we could add it to this PR.